### DISCUSSION TITLE
IP주소체계

### DISCUSSION THREAD

**Discussion by jindream6128** [24tngus/CS_STUDY#37] [Network]
# IP주소, MAC 주소, ARP, RARP

`IP 주소`  (Internet Protocol address) : 논리적 주소, 컴퓨터 네트워크 장치들이 서로를 인식하는 주소로서 변할 수 있다.

`MAC 주소` (Media Access Control Address) : 물리적 주소, 네트워크 인터페이스에 할당 된 고유 식별자 보통 NIC(Network Interface Controller) 에 할당된다.

→ 48비트 구성 : 24비트 OUI(IEEE에서 할당한 제조사 코드) + 24비트 UAA (제조사에서 구별되는 코드)

→ 일반적으로 MAC 주소는 유일하지만, 그렇지 않을 때에는 동일한 네트워크에서만 중복되지 않으면 문제되지 않는다 !

`ARP` (Address Resolution Protocol):  

MAC 주소는 ARP를 통해 파악이 가능하다.

ARP는 논리적 주소인 IP 주소를 물리적 주소인 MAC 주소로 변환해 준다. 

RARP 는 반대인 과정

![1](https://github.com/24tngus/CS_STUDY/assets/101094583/69acdfe5-5ebe-47b4-913c-c1db9f3e520f)

 → 해당 IP주소에 맞는 MAC 주소를 찾기 위해 해당 데이터를 “브로드캐스팅” 을 통해 연결된 네트워크에 있는 장치들에게 모두 보냅니다.

맞는 장치가 있다면 해당 장치는 보낸 장치에게 유니캐스트로 데이터를 전달해 주소를 찾게된다.

![2](https://github.com/24tngus/CS_STUDY/assets/101094583/7a1894ba-d2cb-4b48-9dfa-e2987611fde8)

# IP주소 체계 IPv4와 IPv6

## IPv4

32비트로 표현되는 주소 체계이며 2^32개의 주소를 표현 할 수 있다.

8비트 단위로 점을 찍어 4개로 구분하며, 이 주소 체계  만으로 부족한 부분을 NAT, 서브네팅 등의 기술로 보완한다.

![3](https://github.com/24tngus/CS_STUDY/assets/101094583/b3e9f069-a232-4d62-9bd1-c64abca57784)

## IPv6

IPv6의 경우 2^128개의 주소를 표현할 수 있어, NAT나 서브네팅과 같은 부수적인 기술이 필요하지 않다.

또한, 16비트씩 8개로 구분하여 콜론(:) 으로 구분하며 앞의 연속되는 0 은 생략할 수 있다.

![4](https://github.com/24tngus/CS_STUDY/assets/101094583/4e186d76-4f74-46b2-9306-4a8730365bdd)

128비트 : 64비트(네트워크 주소) + 64비트(인터페이스 주소)

![5](https://github.com/24tngus/CS_STUDY/assets/101094583/4fcc0362-f0f8-41da-a427-6a2015ca0fc5)

### 장점

- IPSec 은 데이터 패킷을 암호화하는 네트워크 프로토콜 제품군인데 IPv6는 포함되어있다
- 단순해진 헤더 포맷으로 IPv4 헤더의 불필요한 필드를 제거하여 보다 빠른 처리가 가능하다.

![6](https://github.com/24tngus/CS_STUDY/assets/101094583/2684a29a-c8b2-4f55-84ce-ffaa8808a489)

- 또한 IPv4는 체크섬이 있지만, IPv6는 체크섬이 없다.
    - **TCP UDP의 체크섬 필드가 있어서 하위의 IPv6로 업그레이드 되면서 사라졌다.**
    
    ![7](https://github.com/24tngus/CS_STUDY/assets/101094583/7323b566-d392-4181-ae34-bbd048c01526)
    
    → 이때 체크섬을 안쓸꺼면 그냥 0으로 만들면 된다. 덕분에 유동적 사용 가능
    
    BUT
    
    IPv6 + UDP 사용시 IPv6의 헤더내의 체크섬 필드가 없기 때문에 무조건 UDP 헤더의 체크섬 필드를 사용한다고 해야한다.
    

## IPv4 vs IPv6 뭐가 더 좋나?

- IPv6는 IPv4 보다 더 많은 주소를 표현할 수 있어 따라서 NAT나, 서브네팅과 같은 부수적인 기술이 필요가 없기때문에 더 좋습니다.
- 또한 IPv4 보다 불필요한 헤더를 제거하여 더 빠르고 기본적으로 IPsec 이라는 네트워크 보안 제품군을 포함하고 있어 보안적 측면에서도 좋습니다.
- 보통은 IPv6 를 사용하는것이 IPv4보다 속도도 빠르고 더 좋지만 일반 사용사례에서는 예외가 있을 수도 있습니다, 현재 네트워크 인프라 및 기기 지원등의 문제등으로 IPv4가 일반적으로 쓰이는 경우 예외가 생길수도 있지만, 장기적인 측면이나, 확장성을 보아서는 IPv6 사용을 적극적으로 고려할 필요가 있습니다.

## IP 주소체계 클래스풀(Classful IP Addressing)

`IP주소`는 : `네트워크 주소` + `호스트 주소`로 나누어진다.

- 네트워크 주소가 동일하다 → 동일한 네트워크
- 호스트 주소: 각각의 호스트를 구분하는 주소

클래스풀 이란 네트워크 주소를 매기고, 그에 따라 네트워크의 크기를 다르게 하여 각각의 클래스를 할당하는 주소 체계, 구분하는 기준자를 서브넷 마스크 라고 합니다.

![8](https://github.com/24tngus/CS_STUDY/assets/101094583/890654b4-fecb-4126-a684-e97b61af4a5b)

## A 클래스

![9](https://github.com/24tngus/CS_STUDY/assets/101094583/38d6e022-2629-4aa3-8356-72e73857e6ca)

- 2^24 - 2 한 네트워크당 약 1600만개의 호스트 아이디
- 네트워크 주소 범위 : `1 ~ 126` , 127.X → 루프백 주소, 0.0.0.0 → 특수주소

## B 클래스

![10](https://github.com/24tngus/CS_STUDY/assets/101094583/45232f9a-e1d9-46fb-8684-c744d016b774)

- 2^16 -2 한 네트워크 당 약 6만 5천개
- 네트워크 주소 범위 : `128 ~ 191`

## C 클래스

![11](https://github.com/24tngus/CS_STUDY/assets/101094583/e71b494a-780e-428c-8b45-f9e8440ceea6)

- 2^8 - 2 한 네트워크당 254개
- 네트워크 주소 범위 : `192~223`

### Why? -2 인가

→ 제일 앞 주소는 네트워크 주소.

→ 제일 마지막 주소는 브로드캐스팅 주소

### 문제점!

- 네트워크 크기가 작은 경우, 큰 네트워크를 필요로 하는 조직은 여러개를 확보해야 하는 어려움,
- 반대인 경우 IP낭비

## IP주소 체계 4 : 클래스 리스, 서브넷 마스크, 서브네팅

→ 클래스풀의 단점을 해결하기 위한 클래스 리스!

→ 클래스로 나누는 것이 나닌 서브넷 마스크를 중심으로 어디까지가 네트워크 주소이고, 어디까지가 호스트 주소인지를 구분하는것. 

`서브네팅` : 네트워크를 나눈다 !

`서브넷` : 서브네트워크, 쪼개진 네트워크

`서브넷마스크` : 서브네트워크를 위한 비트마스크

![12](https://github.com/24tngus/CS_STUDY/assets/101094583/0edbd3d3-f337-4ab7-be73-0ea90019a312)

255 → 11111111(2)

T(1) * T(1) → T(1)

T(1) * F(0) → F(0)

F(0) * T(1) → F(0)

F(0) * F(0) → F(0)

## IP주소 체계 5 : 공인IP, 사설 IP NAT

IP 주소의 부족을 공인 IP와 사설 IP로 나누고 중간에는 NAT 이라는 기술을 통해 해결합니다.

`NAT`(Network Address Translation) 는 패킷이 트래픽 라우팅 장치를 통해 전송되는 동안 패킷의 IP주소를 변경, IP주소를 다른 IP 주소로 매핑하는 방법. → NAT 을 통해 내부 네트워크 노출되지 않는다는 장점이있다.

![13](https://github.com/24tngus/CS_STUDY/assets/101094583/1665aea4-6acc-4279-b745-15dbd5d28d93)

공유기 ↔ NAT

실 생활에서 인터넷 회선 하나를 개통하고 보통 공유기를 써서 wifi를 만드는데 이때 여러 대의 호스트가 하나의 공인 IP 주소를 사용하여 인터넷에 접속하게 된다.

**Comment by 24tngus** [comment]
확인했습니다

**Comment by berryberries** [comment]
확인했습니다!
